src/mo_frontend/typing.ml

```diff
@@ -1529,12 +1532,13 @@ and infer_exp'' env exp : T.typ =
     if not env.pre then
       coverage_cases "switch" env cases t1 exp.at;
     t
-  | TryE (exp1, cases) ->
+  | TryE (exp1, cases, exp2_opt) ->
     let t1 = infer_exp env exp1 in
     let t2 = infer_cases env T.catch T.Non cases in
     if not env.pre then begin
+      Option.iter (check_exp_strong { env with rets = None; labs = T.Env.empty } T.unit) exp2_opt;
```

Comment:
Nice - no returns no exiting jumps allowed. How are throws prevented though?

Reply:
```suggestion
      Option.iter (check_exp_strong { env with rets = None; labs = T.Env.empty; throws = None } T.unit) exp2_opt;
```
Maybe add a test that incorrectly tries to throw from finally and then make this change.

Reply:
`throws are prevented by the effect system already (see `test/fail/try-finally`). But this is a good defensive measure.

Reply:
See #4622 for the fix